File: Solutions/1140. Stone Game II/1140. Stone Game II.cpp
class Solution {
public:
    int stoneGameII(vector<int>& piles) {
        int n=piles.size();
        vector<vector<int>> dp(n+1,vector<int>(n+1));
        vector<int> sufsum(n+1);
        for(int i=n-1;i>=0;--i)
            sufsum[i]=sufsum[i+1]+piles[i];
        for(int i=0;i<=n;++i)
            dp[i][n]=sufsum[i];
        for(int i=n-1;i>=0;--i)
            for(int j=n-1;j>=1;--j)
                for(int X=1;X<=2*j and i+X<=n;X++)
                    dp[i][j]=max(dp[i][j],sufsum[i]-dp[i+X][max(j,X)]);
        return dp[0][1];
    }
};

File: Solutions/1140. Stone Game II/1140. Stone Game II.py
class Solution:
    def stoneGameII(self, piles: List[int]) -> int:
        n=len(piles)
        dp=[[0]*(n+1) for _ in range(n+1)]
        sufsum=[0]*(n+1)
        for i in range(n-1,-1,-1):
            sufsum[i]=sufsum[i+1]+piles[i]
        for i in range(n+1):
            dp[i][n]=sufsum[i]
        for i in range(n-1,-1,-1):
            for j in range(n-1,0,-1):
                X=1
                while X<=2*j and i+X<=n:
                    dp[i][j]=max(dp[i][j],sufsum[i]-dp[i+X][max(j,X)])
                    X+=1
        return dp[0][1]